### PR TITLE
add DeprecatedObjectReference in Destination to support backwards compatibility

### PR DESCRIPTION
**Pull request by capri-xiyue**
Regarding discussion of https://github.com/knative/eventing/issues/1918, to avoid the breaking api change for sources, we can add DeprecatedObjectReference in Destination to support backwards compatibility. This is a suggestion change. We have not made a decision yet.
## Proposed Changes

- added DeprecatedObjectReference in Destination to support backwards compatible
- Added unit tests for new business logic

**Release Note**

```release-note
This is a suggestion change. We have not made a decision yet. Please don't merge it now.
```

### CODE REVIEW

File: apis/v1alpha1/destination.go [hunk 1]
```diff
@@ -28,6 +28,9 @@ type Destination struct {
 	// +optional
 	Ref *corev1.ObjectReference `json:"ref,omitempty"`
 
+	// +optional
+	DeprecatedObjectReference *corev1.ObjectReference `json:",inline"`
```
Comment (n3wscott) [inline apis/v1alpha1/destination.go]: with this change, can you add some json marshaling tests to confirm we can still support both formats?

File: apis/v1alpha1/destination_test.go [hunk 2]
```diff
@@ -80,6 +80,42 @@ func TestValidateDestination(t *testing.T) {
 			},
 			want: "missing field(s): ref.kind",
 		},
+		"valid DeprecatedObjectReference": {
```
Comment (grantr) [inline apis/v1alpha1/destination_test.go]: This test description is incorrect: DeprecatedObjectReference is invalid (missing `kind`)

File: apis/v1alpha1/destination.go (hunk 1 above)
Comment (capri-xiyue) [reply apis/v1alpha1/destination.go]: Do you mean writing unit tests which can accept parameters like  Kubernetes YAML and we verify we can still convert it to a Destination object?

File: apis/v1alpha1/destination_test.go (hunk 2 above)
Comment (n3wscott) [reply apis/v1alpha1/destination_test.go]: There are several: 

https://github.com/knative/pkg/blob/d5915547d9db02c20f3dd1a16a1f62698ea43781/apis/volatile_time_test.go#L53-L74

https://github.com/knative/pkg/blob/ea367c13422f751fac001387b654e087ef0962a2/apis/url_test.go#L89-L125

https://github.com/knative/pkg/blob/281cda84ceb316a70c2eafc5b3250a4b72c2d468/apis/condition_types_test.go#L155-L224

**Comment by n3wscott** [comment]
> @n3wscott do you have an example of similar tests?

There are several: 

https://github.com/knative/pkg/blob/d5915547d9db02c20f3dd1a16a1f62698ea43781/apis/volatile_time_test.go#L53-L74

https://github.com/knative/pkg/blob/ea367c13422f751fac001387b654e087ef0962a2/apis/url_test.go#L89-L125

https://github.com/knative/pkg/blob/281cda84ceb316a70c2eafc5b3250a4b72c2d468/apis/condition_types_test.go#L155-L224

**Comment by grantr** [comment]
@n3wscott do you want to create an issue tracking the addition of json marshaling tests?

**Comment by capri-xiyue** [comment]
> @n3wscott do you want to create an issue tracking the addition of json marshaling tests?

I will create the issue and add marshal tests. I have the marshal tests in my local. I didn't push it for this PR.

**Pull request merged**